server/lib/src/idm/oauth2.rs

```diff
@@ -2701,12 +2815,38 @@ fn validate_scopes(req_scopes: &BTreeSet<String>) -> Result<(), Oauth2Error> {
     Ok(())
 }
 
+/// device code is a random bucket of bytes used in the device flow
+#[inline]
+fn gen_device_code() -> Result<[u8; 16], Oauth2Error> {
+    let mut rng = rand::thread_rng();
+    let mut result = [0u8; 16];
+    if let Err(err) = rng.try_fill(&mut result) {
+        error!("Failed to generate device code! {:?}", err);
+        return Err(Oauth2Error::ServerError(OperationError::Backend));
+    }
+    Ok(result)
+}
+
+#[inline]
+/// Returns (xxx-yyy-zzz, digits) where one's the human-facing code, the other is what we store in the DB.
+fn gen_user_code() -> (String, u32) {
+    use rand::Rng;
+    let mut rng = rand::thread_rng();
+    let num: u32 = rng.gen_range(0..=999999999);
+    let result = format!("{:09}", num);
+    (
+        format!("{}-{}-{}", &result[0..3], &result[3..6], &result[6..9]),
+        num,
+    )
+}
+
 #[cfg(test)]
 mod tests {
```

Comment:
Still no tests here?

Reply:
This is where the bulk of the logic and correctness testing needs to be, even if they are "failing" initially.

Reply:
There's e2e tests being built in testkit https://github.com/yaleman/kanidm/blob/754114d4fc9ac6c18fb59c0dc1c998a7a8ac05bd/server/testkit/tests/oauth2_device_flow.rs

Reply:
(But this is still very much a draft currently)

Reply:
Normally the testkit tests are to check the api works "as we expect" but we put all the detailed logic tests in here because we can control parameters like time and replication for example :)

Reply:
I will be much more able to add them once we've figured out what the internals will look like, and there's something to test, but will add some stuff before this un-drafts 😄

Reply:
Yeah, I think I just want to be clear the majority of the testing should be here, not in testkit. Testkit is just for api checking etc.